book-content/chapters/05-unions-literals-and-narrowing.md

```diff
@@ -0,0 +1,1856 @@
+
+```ts twoslash
+// @errors: 2345
+type PhysicalFormat = "LP" | "CD" | "Cassette";
+// ---cut---
+const getAlbumFormats = (format: PhysicalFormat) => {
+  // function body
+};
+
+getAlbumFormats("MP3");
+```
+
+### Exercises
+
+#### Exercise 1: `string` or `null`
+
+Here we have a function called `getUsername` that takes in a `username` string. If the `username` is not equal to `null`, we return a new interpolated string. Otherwise, we return `"Guest"`:
+
+```typescript
+function getUsername(username: string) {
+  if (username !== null) {
+    return `User: ${username}`;
+  } else {
+    return "Guest";
+  }
+}
+```
+
+In the first test, we call `getUsername` and pass in a string of "Alice" which passes as expected. However, in the second test, we have a red squiggly line under `null` when passing it into `getUsername`:
+
+```ts twoslash
+// @errors: 2345
+import { Equal, Expect } from "@total-typescript/helpers";
+
+function getUsername(username: string) {
+  if (username !== null) {
+    return `User: ${username}`;
+  } else {
+    return "Guest";
+  }
+}
+
+// ---cut---
+const result = getUsername("Alice");
+
+type test = Expect<Equal<typeof result, string>>;
+
+const result2 = getUsername(null);
+
+type test2 = Expect<Equal<typeof result2, string>>;
+```
+
+Normally we wouldn't explicitly call the `getUsername` function with `null`, but in this case it's important that we handle `null` values. For example, we might be getting the `username` from a user record in a database, and the user might or might not have a name depending on how they signed up.
+
+Currently, the `username` parameter only accepts a `string` type, and the check for `null` isn't doing anything. Update the function parameter's type so the errors are resolved and the function can handle `null` .
+
+<Exercise title="Exercise 1: `string` or `null`" filePath="/src/018-unions-and-narrowing/053-introduction-to-unions.problem.ts"></Exercise>
+
+#### Exercise 2: Restricting Function Parameters
+
+Here we have a `move` function that takes in a `direction` of type string, and a `distance` of type number:
+
+```tsx
+function move(direction: string, distance: number) {
+  // Move the specified distance in the given direction
+}
+```
+
+The implementation of the function is not important, but the idea is that we want to be able to move either up, down, left, or right.
+
+Here's what calling the `move` function might look like:
+
+```typescript
+move("up", 10);
+
+move("left", 5);
+```
+
+To test this function, we have some `@ts-expect-error` directives that tell TypeScript we expect the following lines to throw an error.
+
+However, since the `move` function currently takes in a `string` for the `direction` parameter, we can pass in any string we want, even if it's not a valid direction. There is also a test where we expect that passing `20` as a distance won't work, but it's being accepted as well.
```

Comment:
I'm confused by the part I removed in this suggestion:
```suggestion
However, since the `move` function currently takes in a `string` for the `direction` parameter, we can pass in any string we want, even if it's not a valid direction.
```

It seems out of the blue here, I'm not even sure how u expect to correct it later and I don't think there is any follow up question/excercise about it later on.